src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,714 @@
+
+  PStatic VisitExpr_(const RefReadNode* op, LetList* ll) final {
+    PStatic r = VisitExpr(op->ref, ll);
+    if (r->pstatic) {
+      PStatic ret = store_.Lookup(&r->pstatic->get<SRefNode>());
+      if (ret) {
+        return ret;
+      }
+    }
+    return NoStatic(RefReadNode::make(r->dynamic));
+  }
+
+  PStatic VisitExpr_(const CallNode* op, LetList* ll) final {
+    PStatic f = VisitExpr(op->op, ll);
+    std::vector<PStatic> x;
+    tvm::Array<Expr> x_dyn;
+    for (const Expr& e : op->args) {
+      PStatic ps = VisitExpr(e, ll);
+      x.push_back(ps);
+      x_dyn.push_back(ps->dynamic);
+    }
+    if (f->pstatic) {
+      return f->pstatic->get<SClosNode>().func(x, op->attrs, op->type_args, ll);
+    } else {
+      store_.Invalidate();
+      return NoStatic(CallNode::make(f->dynamic, x_dyn, op->attrs, op->type_args));
+    }
+  }
+
+  PStatic VisitExpr_(const FunctionNode* op, LetList* ll) final {
+    if (op->IsPrimitive()) {
+      return HasStatic(SClos(ConstEvaluateClos(GetRef<Expr>(op), ll)), GetRef<Expr>(op));
+    } else {
+      Clos f = [=](const std::vector<PStatic>& pv,
+                   const Attrs& attrs,
+                   const tvm::Array<Type>& type_args,
+                   LetList* ll) {
+        return env_.Extend<PStatic>([&]() {
+            CHECK_EQ(pv.size(), op->params.size());
+            for (size_t i = 0; i < pv.size(); ++i) {
+              env_.Insert(op->params[i], pv[i]);
+            }
+            tvm::Map<TypeVar, Type> subst;
+            for (size_t i = 0; i < type_args.size(); ++i) {
+              subst.Set(op->type_params[i], type_args[i]);
+            }
+            for (size_t i = type_args.size(); i < op->type_params.size(); ++i) {
+              subst.Set(op->type_params[i], Type());
+            }
+            return VisitExpr(TypeSubst(op->body, subst), ll);
+          });
+      };
+      Expr dyn = store_.Extend<Expr>([&]() {
+          return FunctionNode::make(op->params, LetList::With([&](LetList* ll) {
+                std::vector<PStatic> pv;
+                for (const auto& v : op->params) {
+                  pv.push_back(NoStatic(v));
+                }
+                tvm::Array<Type> type_args;
+                for (const auto& tp : op->type_params) {
+                  type_args.push_back(tp);
+                }
+                return f(pv, Attrs(), type_args, ll)->dynamic;
+              }), op->ret_type, op->type_params, op->attrs);
+        });
+      return HasStatic(SClos(f), dyn);
+    }
+  }
+
+  Expr Reflect(const PStatic& st) {
+    if (const STensorNode* op = st->pstatic->try_get<STensorNode>()) {
+      return ConstantNode::make(op->data);
+    } else if (const STupleNode* op = st->pstatic->try_get<STupleNode>()) {
+      tvm::Array<Expr> fields;
+      for (const PStatic& field : op->fields) {
+        fields.push_back(Reflect(field));
+      }
+      return TupleNode::make(fields);
+    } else {
+      LOG(ERROR) << "Unknown case";
+      throw;
```

Comment:
is this error actually recoverable (and not evidence of a bogus program)? if not `LOG(FATAL) << "Unknown case";` might be more appropriate.